From c9a65d4ae7e810b47b09f24feb7de0f3db74b54e Mon Sep 17 00:00:00 2001
From: Michael Goffioul <michael.goffioul@gmail.com>
Date: Wed, 28 Sep 2022 08:13:21 +0200
Subject: [PATCH 26/28] Add media extensions for FFMPEG codec2 service

A plain buffer is used to transfer extra information from extractor to
codecs. The buffer is mapped to a custom C2 paramkey.

Backported to r-x86 by hmtheboy154

Change-Id: Id9e36ae90b194b2973ed16af07a9e2b5230aaab7
---
 media/codec2/core/include/C2Config.h                       | 7 +++++++
 media/codec2/sfplugin/CCodecConfig.cpp                     | 5 +++++
 media/libstagefright/FFMPEGUtil.cpp                        | 1 +
 media/libstagefright/Utils.cpp                             | 1 +
 .../include/media/stagefright/MetaDataBase.h               | 1 +
 5 files changed, 15 insertions(+)

diff --git a/media/codec2/core/include/C2Config.h b/media/codec2/core/include/C2Config.h
index 29bccd5dc5..1ea5ef2b4a 100644
--- a/media/codec2/core/include/C2Config.h
+++ b/media/codec2/core/include/C2Config.h
@@ -249,6 +249,8 @@ enum C2ParamIndexKind : C2Param::type_index_t {
 
     // low latency mode
     kParamIndexLowLatencyMode, // bool
+    // FFMPEG
+    kParamIndexRawCodecData,
 };
 
 }
@@ -2182,6 +2184,11 @@ inline C2TimestampGapAdjustmentStruct::C2TimestampGapAdjustmentStruct()
 typedef C2PortParam<C2Tuning, C2TimestampGapAdjustmentStruct> C2PortTimestampGapTuning;
 constexpr char C2_PARAMKEY_INPUT_SURFACE_TIMESTAMP_ADJUSTMENT[] = "input-surface.timestamp-adjustment";
 
+// FFMEG
+
+typedef C2StreamParam<C2Info, C2BlobValue, kParamIndexRawCodecData> C2StreamRawCodecDataInfo;
+constexpr char C2_PARAMKEY_RAW_CODEC_DATA[] = "coded.raw-codec-data";
+
 /// @}
 
 #endif  // C2CONFIG_H_
diff --git a/media/codec2/sfplugin/CCodecConfig.cpp b/media/codec2/sfplugin/CCodecConfig.cpp
index 96f86e8d59..23548ba353 100644
--- a/media/codec2/sfplugin/CCodecConfig.cpp
+++ b/media/codec2/sfplugin/CCodecConfig.cpp
@@ -917,6 +917,11 @@ void CCodecConfig::initializeStandardParams() {
     KEY_AUDIO_SESSION_ID // we use "audio-hw-sync"
     KEY_OUTPUT_REORDER_DEPTH
     */
+
+    // FFMPEG
+
+    add(ConfigMapper("raw-codec-data", C2_PARAMKEY_RAW_CODEC_DATA, "value")
+        .limitTo(D::DECODER & D::CODED));
 }
 
 status_t CCodecConfig::initialize(
diff --git a/media/libstagefright/FFMPEGUtil.cpp b/media/libstagefright/FFMPEGUtil.cpp
index 76a7658520..1958196b83 100644
--- a/media/libstagefright/FFMPEGUtil.cpp
+++ b/media/libstagefright/FFMPEGUtil.cpp
@@ -46,6 +46,7 @@ static const MetaKeyEntry MetaKeyTable[] {
    {kKeyCodecId              , "codec-id"               , INT32},
    {kKeyCodedSampleBits      , "coded-sample-bits"      , INT32},
    {kKeyFileFormat           , "file-format"            , INT32},
+   {kKeyRawCodecData         , "raw-codec-data"         , DATA},
    {kKeyRawCodecSpecificData , "raw-codec-specific-data", CSD},
    {kKeyPcmEncoding          , "pcm-encoding"           , INT32},
    {kKeyRVVersion            , "rv-version"             , INT32},
diff --git a/media/libstagefright/Utils.cpp b/media/libstagefright/Utils.cpp
index 463021e5ab..ff85acb088 100644
--- a/media/libstagefright/Utils.cpp
+++ b/media/libstagefright/Utils.cpp
@@ -789,6 +789,7 @@ static std::vector<std::pair<const char *, uint32_t>> bufferMappings {
         { "text-format-data", kKeyTextFormatData },
         { "thumbnail-csd-hevc", kKeyThumbnailHVCC },
         // FFMEG
+        { "raw-codec-data", kKeyRawCodecData },
         { "raw-codec-specific-data", kKeyRawCodecSpecificData },
     }
 };
diff --git a/media/libstagefright/include/media/stagefright/MetaDataBase.h b/media/libstagefright/include/media/stagefright/MetaDataBase.h
index c6da9c0dae..f4a8771677 100644
--- a/media/libstagefright/include/media/stagefright/MetaDataBase.h
+++ b/media/libstagefright/include/media/stagefright/MetaDataBase.h
@@ -140,6 +140,7 @@ enum {
 
     kKeyIsUnreadable      = 'unre',  // bool (int32_t)
 
+    kKeyRawCodecData         = 'rcdt',
     kKeyRawCodecSpecificData = 'rcsd',  // raw data - added to support mmParser
     kKeyDivXVersion       = 'DivX',  // int32_t
     kKeyDivXDrm           = 'QDrm',  // void *
-- 
2.38.1

